File: scripts/import-policies.js
/**
 * Import Role Policies Script
 *
 * Imports comprehensive role-based policies and permissions into Directus.
 * Based on complete-role-policies.json template.
 */

import { createDirectus, rest, createItem, createItems, readItems } from '@directus/sdk';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Directus connection
const directus = createDirectus(process.env.PUBLIC_URL || 'http://localhost:8055')
  .with(rest());

// Read policies JSON
const policiesPath = path.join(__dirname, '../schema/policies/complete-role-policies.json');
const policiesData = JSON.parse(fs.readFileSync(policiesPath, 'utf8'));

console.log('📋 Importing Role-Based Access Control Policies...\n');

async function importPolicies() {
  try {
    // Login as admin
    console.log('🔐 Authenticating as admin...');

    const authResponse = await fetch(`${process.env.PUBLIC_URL || 'http://localhost:8055'}/auth/login`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        email: process.env.ADMIN_EMAIL || 'admin@dev.local',
        password: process.env.ADMIN_PASSWORD || 'DevPassword123!'
      })
    });

    if (!authResponse.ok) {
      throw new Error('Authentication failed');
    }

    const { data: { access_token } } = await authResponse.json();
    console.log('✅ Authenticated\n');

    // Helper function to make authenticated requests
    const fetchAuth = async (endpoint, options = {}) => {
      const response = await fetch(`${process.env.PUBLIC_URL || 'http://localhost:8055'}${endpoint}`, {
        ...options,
        headers: {
          'Authorization': `Bearer ${access_token}`,
          'Content-Type': 'application/json',
          ...options.headers
        }
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`API Error: ${response.status} - ${error}`);
      }

      return response.json();
    };

    // Get existing policies to check for duplicates
    console.log('🔍 Checking existing policies...');
    const { data: existingPolicies } = await fetchAuth('/policies');
    const existingPolicyNames = new Set(existingPolicies.map(p => p.name));

    // Create policies
    console.log('📦 Creating policies...\n');
    const policyIdMap = {}; // Map template IDs to actual UUIDs

    for (const policy of policiesData.policies) {
      if (existingPolicyNames.has(policy.name)) {
        console.log(`   ⚠️  Policy "${policy.name}" already exists, skipping...`);
        // Get existing policy ID
        const existing = existingPolicies.find(p => p.name === policy.name);
        policyIdMap[policy.id] = existing.id;
        continue;
      }

      console.log(`   ➕ Creating policy: ${policy.name}`);
      const { id, ...policyData } = policy; // Remove template ID

      const { data: created } = await fetchAuth('/policies', {
        method: 'POST',
        body: JSON.stringify(policyData)
      });

      policyIdMap[policy.id] = created.id; // Map template ID to real UUID
      console.log(`      ✅ Created with ID: ${created.id}`);
    }

    console.log('\n✅ All policies created\n');

    // Create permissions for each policy
    console.log('🔐 Creating permissions...\n');
    let permissionCount = 0;

    for (const permission of policiesData.permissions) {
      // Skip comment entries
      if (permission.comment) {
        console.log(`\n${permission.comment}`);
        continue;
      }

      // Replace template policy ID with real UUID
      const permissionData = {
        ...permission,
        policy: policyIdMap[permission.policy]
      };

      try {
        await fetchAuth('/permissions', {
          method: 'POST',
          body: JSON.stringify(permissionData)
        });

        permissionCount++;
        const policyName = policiesData.policies.find(p => p.id === permission.policy)?.name || permission.policy;
        console.log(`   ✅ ${permission.collection}.${permission.action} (${policyName})`);
      } catch (error) {
        console.log(`   ⚠️  Failed: ${permission.collection}.${permission.action} - ${error.message}`);
      }
    }

    console.log(`\n✅ Created ${permissionCount} permissions\n`);

    // Summary
    console.log('📊 Import Summary:');
    console.log(`   Policies: ${Object.keys(policyIdMap).length}`);
    console.log(`   Permissions: ${permissionCount}`);
    console.log('\n🎉 Import complete! Now link policies to roles via Admin UI or update roles.\n');

    console.log('Next steps:');
    console.log('1. Go to Settings → Roles & Permissions');
    console.log('2. For each role, assign the corresponding policy');
    console.log('3. Test with test users from Issue #23\n');

  } catch (error) {
    console.error('❌ Error:', error.message);
    console.error(error.stack);
    process.exit(1);
  }
}

// Run import
importPolicies();

File: scripts/import-policies.py
#!/usr/bin/env python3
"""
Import Role Policies Script

Imports comprehensive role-based policies and permissions into Directus.
Handles authentication properly without bash escaping issues.
"""

import json
import os
import sys
import requests
from pathlib import Path

# Configuration
API_URL = os.getenv('PUBLIC_URL', 'http://localhost:8055')
ADMIN_EMAIL = os.getenv('ADMIN_EMAIL', 'admin@dev.local')
ADMIN_PASSWORD = os.getenv('ADMIN_PASSWORD', 'DevPassword123!')

# Paths
SCRIPT_DIR = Path(__file__).parent
POLICIES_FILE = SCRIPT_DIR / '../schema/policies/complete-role-policies.json'

print('📋 Importing Role-Based Access Control Policies...\n')

def authenticate():
    """Authenticate and get access token"""
    print('🔐 Authenticating as admin...')

    response = requests.post(
        f'{API_URL}/auth/login',
        json={
            'email': ADMIN_EMAIL,
            'password': ADMIN_PASSWORD
        },
        headers={'Content-Type': 'application/json'}
    )

    if not response.ok:
        print(f'❌ Authentication failed: {response.status_code}')
        print(response.text)
        sys.exit(1)

    data = response.json()
    token = data['data']['access_token']
    print('✅ Authenticated\n')
    return token

def create_policy(token, policy_data):
    """Create a policy and return its ID"""
    name = policy_data['name']
    print(f'   ➕ Creating policy: {name}')

    # Remove template ID
    policy_payload = {k: v for k, v in policy_data.items() if k != 'id'}

    response = requests.post(
        f'{API_URL}/policies',
        json=policy_payload,
        headers={
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }
    )

    if response.ok:
        policy_id = response.json()['data']['id']
        print(f'      ✅ Created with ID: {policy_id}')
        return policy_id
    else:
        # Check if already exists
        existing_response = requests.get(
            f'{API_URL}/policies',
            params={'filter[name][_eq]': name},
            headers={'Authorization': f'Bearer {token}'}
        )

        if existing_response.ok:
            existing_data = existing_response.json()['data']
            if existing_data:
                policy_id = existing_data[0]['id']
                print(f'      ⚠️  Policy already exists with ID: {policy_id}')
                return policy_id

        print(f'      ❌ Failed to create policy: {response.text}')
        return None

def create_permission(token, permission_data):
    """Create a permission"""
    collection = permission_data['collection']
    action = permission_data['action']

    response = requests.post(
        f'{API_URL}/permissions',
        json=permission_data,
        headers={
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }
    )

    if response.ok or response.status_code == 204:
        return True
    else:
        print(f'   ⚠️  Failed: {collection}.{action} - {response.text}')
        return False

def main():
    # Load policies JSON
    print('📂 Loading policies from JSON...')
    with open(POLICIES_FILE) as f:
        data = json.load(f)

    print(f'   Found {len(data["policies"])} policies and {len(data["permissions"])} permissions\n')

    # Authenticate
    token = authenticate()

    # Create policies and map IDs
    print('📦 Creating policies...\n')
    policy_id_map = {}

    for policy in data['policies']:
        template_id = policy['id']
        real_id = create_policy(token, policy)
        if real_id:
            policy_id_map[template_id] = real_id

    print(f'\n✅ Created/found {len(policy_id_map)} policies\n')

    # Create permissions
    print('🔐 Creating permissions...\n')
    permission_count = 0
    failed_count = 0

    for permission in data['permissions']:
        # Skip comment entries
        if 'comment' in permission:
            print(f'\n{permission["comment"]}')
            continue

        # Replace template policy ID with real UUID
        permission_data = permission.copy()
        template_policy_id = permission['policy']

        if template_policy_id in policy_id_map:
            permission_data['policy'] = policy_id_map[template_policy_id]

            if create_permission(token, permission_data):
                permission_count += 1
                policy_name = next((p['name'] for p in data['policies'] if p['id'] == template_policy_id), template_policy_id)
                print(f'   ✅ {permission["collection"]}.{permission["action"]} ({policy_name})')
            else:
                failed_count += 1
        else:
            print(f'   ⚠️  Skipping {permission["collection"]}.{permission["action"]} - policy not found')
            failed_count += 1

    print(f'\n✅ Created {permission_count} permissions')
    if failed_count > 0:
        print(f'⚠️  Failed to create {failed_count} permissions\n')

    # Summary
    print('\n📊 Import Summary:')
    print(f'   Policies: {len(policy_id_map)}')
    print(f'   Permissions: {permission_count}')
    print('\n🎉 Import complete!\n')

    print('Next steps:')
    print('1. Go to Settings → Roles & Permissions')
    print('2. For each role, assign the corresponding policy via the "Access" tab')
    print('3. Test with test users from Issue #23\n')

if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        print(f'\n❌ Error: {e}')
        import traceback
        traceback.print_exc()
        sys.exit(1)
